cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -506,32 +545,44 @@ public void createScheduleTask(Id.Program programId, SchedulableProgramType prog
         try {
           streamSize = pollOnce();
         } catch (IOException e) {
-          // Polling should not fail - the whole logic of stream size schedules relies on it
-          throw new SchedulerException("Polling could not be performed on stream " + streamSizeSchedule.getStreamName(),
-                                       e);
+          if (persist) {
+            // Polling should not fail when creating a schedule for the first time -
+            // the whole logic of stream size schedules relies on it
+            throw new SchedulerException("Polling could not be performed on stream " +
+                                           streamSizeSchedule.getStreamName(),
+                                         e);
+          }
         }
 
         // Initialize the schedule task
-        if (basePollSize == -1 && basePollTs == -1) {
-          newTask.startSchedule(streamSize.getSize(), streamSize.getTimestamp(), lastRunSize, lastRunTs,
-                                active, persist);
+        StreamSizeScheduleTask newTask = new StreamSizeScheduleTask(programId, programType, streamSizeSchedule);
+        if (persist) {
+          // First time that we create this schedule, it has to be initialized with the latest polling info
+          newTask.startSchedule(streamSize.getSize(), streamSize.getTimestamp(),
+                                lastRunSize, lastRunTs, active, true);
         } else {
-          newTask.startSchedule(basePollSize, basePollTs, lastRunSize, lastRunTs, active, persist);
+          newTask.startSchedule(basePollSize, basePollTs, lastRunSize, lastRunTs, active, false);
         }
 
+        // We only modify the scheduleTasks if the persistence in startSchedule() did not throw any exception
+        scheduleTasks.put(scheduleId, newTask);
+
         if (active) {
           activeTasks.incrementAndGet();
         }
       }
 
```

Comment:
I am not sure whether I like using the same method for creating a new schedule and for restoring the state after a restart of the scheduler. In case of a restart, you can simply read all schedules from the store once, the pollOnce and send to all active tasks. If you separate this out into separate methods, it becomes much easier to understand.